### PR TITLE
ALTAPPS-682: Shared current profile state repository

### PR DESCRIPTION
**Pull request by XanderZhu**
**YouTrack Issues**:
[#ALTAPPS-682](https://vyahhi.myjetbrains.com/youtrack/issue/ALTAPPS-682)

**Checklist**

_Before Code Review:_

- [x] Fields "Assignees, Labels, Milestone" are filled in the pull request;
- [x] All checks have been passed;
- [x] Changes have been checked locally.

**Description**

1. Implement `CurrentProfileRepository`
2. Replace `ProfileInteractor.getCurrentProfile` with `CurrentProfileRepository.getState`
3. Replace `ProfileInteractor.observeHypercoinsBalance` & `ProfileInteractor.notifyHypercoinsBalanceChanged` with the `CurrentProfileStateRepository.observeHypercoinsBalance` extension function
4. Replace `ProfileInteractor.selectTrack` with `ProfileRepository.selectTrack`

### CODE REVIEW

**Comment by XanderZhu** [comment]
@ivan-magda @vladkash 
This is a draft of my work, please take a look. Your feedback is welcomed!
I suggest moving `ProfileInteractor.selectTrackWithProject` & `ProfileInteractor.selectTrack` to `ProjectSelectionInteractor` & `TrackSelectionInteractor` respectively because it has nothing to common with the Profile entity but API routes.
Also in the context of this task, I would like to move solvedStepsSharedFlow out of the `ProfileInteractor` to remove it.

File: shared/src/commonMain/kotlin/org/hyperskill/app/profile/domain/repository/CurrentProfileStateRepository.kt
Comment (vladkash) [inline shared/src/commonMain/kotlin/org/hyperskill/app/profile/domain/repository/CurrentProfileStateRepository.kt:14]: I think it's better to change order of `distinctUntilChanged` and `map` methods to emit changes only when `hypercoinsBalance` is changed, in current implementation it will be emitted for example, when `trackId` is changed. Or I don't understand `distinctUntilChanged` method?

Comment (XanderZhu) [reply shared/src/commonMain/kotlin/org/hyperskill/app/profile/domain/repository/CurrentProfileStateRepository.kt:14]: @vladkash 
Yes, you are right, the value will be produced when any of the class fields are changed.
It's better to change the order.

**Review by vladkash — APPROVED** [review]
@XanderZhu please, wait for review of @ivan-magda before merge

File: shared/src/commonMain/kotlin/org/hyperskill/app/profile/presentation/ProfileReducer.kt
Comment (ivan-magda) [inline shared/src/commonMain/kotlin/org/hyperskill/app/profile/presentation/ProfileReducer.kt:1]: Nice, with that changes we are always keeping a profile up to date 👍

**Review by ivan-magda — COMMENTED** [review]
On application launch, I'm seeing 2 multiple remote requests for the `/api/profiles/current`. I suppose one comes from `HomeFeature` and the other from `GamificationToolbarFeature`.

Screenshot 2023-06-09 at 2 10 14 PM


I'm thinking that we need to force load from the remote only from root features like Home and embedded features should load with a primary source of cache or observe changes of the current profile, as you do in `ProfileActionDispatcher`:
```kotlin
currentProfileStateRepository.changes
    .distinctUntilChanged()
    .onEach { profile ->
        onNewMessage(Message.ProfileChanged(profile))
    }
    .launchIn(actionScope)
```

**This is just a note, out of the scope of this task. I think this is another `debt` task with improvements.**

File: shared/src/commonMain/kotlin/org/hyperskill/app/main/domain/interactor/AppInteractor.kt
Comment (ivan-magda) [reply shared/src/commonMain/kotlin/org/hyperskill/app/main/domain/interactor/AppInteractor.kt:30]: Related comment https://github.com/hyperskill/mobile-app/pull/519#discussion_r1223844036

File: shared/src/commonMain/kotlin/org/hyperskill/app/profile/cache/CurrentProfileStateHolderImpl.kt
Comment (XanderZhu) [reply shared/src/commonMain/kotlin/org/hyperskill/app/profile/cache/CurrentProfileStateHolderImpl.kt:13]: @ivan-magda 
``` kotlin
when {
    else -> {
        return null
    }
}
```
`return null` in this context means `return the whole function result`, not `return when-statement result`.
You don't need any keyword to return the result from the `when-statement`.

Comment (XanderZhu) [reply shared/src/commonMain/kotlin/org/hyperskill/app/profile/cache/CurrentProfileStateHolderImpl.kt:13]: So, the compiler is right, and the `key` variable type is truly String.

File: shared/src/commonMain/kotlin/org/hyperskill/app/core/domain/repository/StateRepository.kt
Comment (XanderZhu) [reply shared/src/commonMain/kotlin/org/hyperskill/app/core/domain/repository/StateRepository.kt:52]: > And if I understood correctly this also will propagate changes and all subscribers will receive a new (changed) profile.

Yes, the new state produced by lambda will be propagated to the `changes` subscribers.

File: shared/src/commonMain/kotlin/org/hyperskill/app/profile/injection/ProfileDataComponent.kt
Comment (ivan-magda) [reply shared/src/commonMain/kotlin/org/hyperskill/app/profile/injection/ProfileDataComponent.kt:9]: OK, then we need to make `CurrentProfileStateRepository` singleton, to make it work correctly.

If I understood correctly we create a new instance each time:

AppGraph
```kotlin
fun buildProfileDataComponent(): ProfileDataComponent
```

ProfileDataComponentImpl
```kotlin
override val currentProfileStateRepository: CurrentProfileStateRepository
    get() = CurrentProfileStateRepositoryImpl(
        profileRemoteDataSource = profileRemoteDataSource,
        stateHolder = currentProfileStateHolder,
        mutableSharedFlow = currentProfileSharedFlow
    )
```

CurrentProfileStateRepositoryImpl -> BaseStateRepository -> mutex

Comment (XanderZhu) [reply shared/src/commonMain/kotlin/org/hyperskill/app/profile/injection/ProfileDataComponent.kt:9]: Good point, I didn't think about the Mutex recreation.

**Comment by XanderZhu** [comment]
@ivan-magda 
Our AppComponents have a crucial problem: singleton components are initialized at creation time (not lazy) and there are singleton components that are dependent on other singleton components and they get this dependency via the `appGraph` link, not directly. In this situation, the order of singleton components initialization in the graph is important and can cause a runtime crash.
This is why I add direct components as dependencies to the `analyticComponent` and `profileDataComponent`.

**Review by ivan-magda — APPROVED** [review]

**Pull request merged**
